### ISSUE TITLE
Feature request: Introduce different states to prevent clusters from being turned inactive programmatically

### ISSUE THREAD

**Issue by andythsu** [trinodb/trino-gateway#80]
Currently, if the backends are not responding to healthcheck, trino-gateway will inactivate the backends. Since healthcheck only checks against healthy backends, inactive backends will never be turned to active again. Therefore, it is better to introduce different states. It is proposed to introduce healthy/pending/unhealthy states depending on the health of the backends. These states will be independent from the active/inactive states, so it ensures that unhealthy but active backends will always be checked and possibly turned active again.

`Healthy state` is defined as backends are healthy and ready to be served
`Pending state` is defined as when the backend is switched from inactive to active. It will wait until healthcheck returns success before turning backend to healthy
`Unhealthy state` is defined as backends are returning error or not responding to healthcheck. At this point, if the backends are still active, healthcheck will still check on these backends.

**Comment by andythsu** [comment]
When backend is active and health check passes, health state should be HEALTHY
![image](https://github.com/trinodb/trino-gateway/assets/18670495/c7c23c93-cf98-4ab2-ba09-bf87cef6826a)

When backend is inactive, health state should be PENDING
![image](https://github.com/trinodb/trino-gateway/assets/18670495/696e0d96-206f-4ad0-b592-767b871ef421)

When backend is turned active from inactive, health state should be PENDING until health check passes
![image](https://github.com/trinodb/trino-gateway/assets/18670495/a2c6b57d-1497-4520-bcba-35557f5a27fd)

**Comment by willmostly** [comment]
Instead of deactivating backends, we could switch to using the `TrinoQueueLengthRoutingTable` as default. This routing strategy will automatically ignore any backend that does not respond with the number of running and queued queries.

**Comment by Chaho12** [comment]
At first I thought gateway only checks against active backends because when user disables backends, e.g for deployment purpose, it should not go back to enabled status automatically. A manual interaction is needed. 

> unhealthy but active backends can still be checked and possibly turned active again.

So, if i understood it correctly, healthy status checks only to those who has `active status as true` right? Those in active as false is not checked? Am just clarifying what you intended :)

**Comment by andythsu** [comment]
> At first I thought gateway only checks against active backends because when user disables backends, e.g for deployment purpose, it should not go back to enabled status automatically. A manual interaction is needed.
> 
> > unhealthy but active backends can still be checked and possibly turned active again.
> 
> So, if i understood it correctly, healthy status checks only to those who has `active status as true` right? Those in active as false is not checked? Am just clarifying what you intended :)

this is correct. `ActiveClusterMonitor` only checks against **active** backends. 

https://github.com/trinodb/trino-gateway/blob/e155380c13b44d158117ede1171f0187dcaf8ce8/gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ActiveClusterMonitor.java#L62

**Comment by andythsu** [comment]
> Instead of deactivating backends, we could switch to using the `TrinoQueueLengthRoutingTable` as default. This routing strategy will automatically ignore any backend that does not respond with the number of running and queued queries.

I think it's more of an UX improvement because deactivating backends also shows backends' health as "unhealthy" on UI. It lets users know that the backends are down at the moment.

Open to any other discussions.

**Comment by rdsarvar** [comment]
> Instead of deactivating backends, we could switch to using the TrinoQueueLengthRoutingTable as default. This routing strategy will automatically ignore any backend that does not respond with the number of running and queued queries.

@willmostly out of curiosity what would this implementation do if the active node count is 0 for a coordinator, would it still route because running == queued == 0?

my vote (albeit not important) is the current implementation you've went with because it doesn't affect routing and fixes (IMO) a broken functionality where backends can't self heal

The only risky part of this implementation that I can see is that a flapping backend can continually rejoin the routing, accept + fail queries, drop as unhealthy, repeat. But I believe the benefit of a healthy backend being able to rejoin routing after a false positive health failure out weights the cons.

**Comment by jeffreychucASAPP** [comment]
We actually implemented this feature on our internal fork of trino-gateway, I'll see about getting it cleared and uploaded here as a branch.  This feature was particularly important for our deployment as we have this deployed on k8s and we'll have redeployments of our trino clusters that would bring them in and out of "active".

**Comment by xkrogen** [comment]
Can we close this now that we have https://github.com/trinodb/trino-gateway/pull/360 merged? cc @mosabua

**Comment by mosabua** [comment]
Not sure .. wdyt @andythsu ... is everything in place for your needs to close this issue?

**Comment by mosabua** [comment]
Also we have UNKNOWN as state as well
